### PR TITLE
feat(aws-codecommit): use CloudWatch Events instead of polling by default in the CodePipeline Action

### PR DESCRIPTION
**Pull request by skinny85**
This is the preferred way now by the CodePipeline team, so we should make it the default.

--------------------------------------------------------

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license.

### CODE REVIEW

File: packages/@aws-cdk/aws-codecommit/lib/pipeline-action.ts [hunk 1]
```diff
@@ -54,11 +54,15 @@ export class PipelineSourceAction extends codepipeline.SourceAction {
       configuration: {
         RepositoryName: props.repository.repositoryName,
         BranchName: props.branch || 'master',
-        PollForSourceChanges: props.pollForSourceChanges !== undefined ? props.pollForSourceChanges : true
+        PollForSourceChanges: props.pollForSourceChanges || false,
       },
       outputArtifactName: props.outputArtifactName
     });
 
+    if (props.pollForSourceChanges !== true) {
```
Comment (eladb) [inline packages/@aws-cdk/aws-codecommit/lib/pipeline-action.ts]: Ideally you wouldn't want to codify the default multiple times in your code.

Since the default is false, I would just write this `if (!props.pollForSourceChanges)`.
Alternatively, if you want to be more formal, create a local variable:

```ts
const pollForSourceChanges = props.pollForSourceChanges || false;
```

And use it everywhere.

File: packages/@aws-cdk/aws-codecommit/lib/pipeline-action.ts [hunk 2]
```diff
@@ -54,11 +54,15 @@ export class PipelineSourceAction extends codepipeline.SourceAction {
...
+      props.repository.onCommit('CodePipeline', props.stage.pipelineAsEventTarget, props.branch || 'master');
```
Comment (eladb) [inline packages/@aws-cdk/aws-codecommit/lib/pipeline-action.ts]: What happens if we want multiple pipelines to subscribe to the same repo? I think your ID should include the pipeline's `.uniqueId` or something.

Comment (eladb) [inline packages/@aws-cdk/aws-codecommit/lib/pipeline-action.ts]: Again, codify the default for `branch` once in your code:

```ts
const branch = props.branch || 'master';
```

File: packages/@aws-cdk/aws-codepipeline/lib/stage.ts [hunk 3]
```diff
@@ -147,6 +147,10 @@ export class Stage extends cdk.Construct implements actions.IStage, actions.IInt
     return this.pipeline.role;
   }
 
+  public get pipelineAsEventTarget(): events.IEventRuleTarget {
```
Comment (eladb) [inline packages/@aws-cdk/aws-codepipeline/lib/stage.ts]: Why is this needed? The caller can just do `stage.pipeline` - it will also be much more readable at the call site.

File: packages/@aws-cdk/aws-codecommit/lib/pipeline-action.ts (hunk 2 above)
Comment (eladb) [inline packages/@aws-cdk/aws-codecommit/lib/pipeline-action.ts]: Why not just `stage.pipeline`?

File: packages/@aws-cdk/aws-codepipeline/test/test.cloudformation-pipeline-actions.ts [hunk 4]
```diff
@@ -30,6 +30,7 @@ export = {
     stage: sourceStage,
     outputArtifactName: 'SourceArtifact',
     repository: repo,
+    pollForSourceChanges: true,
   });
 
   /** Build! */
```
Comment (eladb) [inline packages/@aws-cdk/aws-codepipeline/test/test.cloudformation-pipeline-actions.ts]: Missing unit test for the default and setting to `false`

File: packages/@aws-cdk/aws-codecommit/lib/pipeline-action.ts (hunk 1 above)
Comment (skinny85) [reply packages/@aws-cdk/aws-codecommit/lib/pipeline-action.ts]: We can't use a variable here, as the call to `super` must be the first statement in the constructor.

Will change to `if (!props.pollForSourceChanges)`.

File: packages/@aws-cdk/aws-codecommit/lib/pipeline-action.ts (hunk 2 above)
Comment (skinny85) [reply packages/@aws-cdk/aws-codecommit/lib/pipeline-action.ts]: Good call. It's actually a miss on my part, as I meant to change this name to something else, but forgot. Will add the Pipeline's `uniqueId`.

Comment (skinny85) [reply packages/@aws-cdk/aws-codecommit/lib/pipeline-action.ts]: Again, can't use a variable here (`super` call, yada yada yada). The only alternative is a function, and that seemed like overkill to me here, hence the duplication.

File: packages/@aws-cdk/aws-codepipeline/lib/stage.ts (hunk 3 above)
Comment (skinny85) [reply packages/@aws-cdk/aws-codepipeline/lib/stage.ts]: A caller cannot do `stage.pipeline`, because you cannot reference the `codepipeline` module from the `codepipeline-api` one (where the `IStage` interface lives).

File: packages/@aws-cdk/aws-codecommit/lib/pipeline-action.ts (hunk 2 above)
Comment (skinny85) [reply packages/@aws-cdk/aws-codecommit/lib/pipeline-action.ts]: See above - it would cause a circular dependency. The entire reason for breaking out the `codepipeline-api` module is because Actions *cannot* reference the `codepipeline` module, where the `Pipeline` type lives.

File: packages/@aws-cdk/aws-codepipeline/test/test.cloudformation-pipeline-actions.ts (hunk 4 above)
Comment (skinny85) [reply packages/@aws-cdk/aws-codepipeline/test/test.cloudformation-pipeline-actions.ts]: Sure.

File: packages/@aws-cdk/aws-codecommit/lib/pipeline-action.ts (hunk 1 above)
Comment (eladb) [reply packages/@aws-cdk/aws-codecommit/lib/pipeline-action.ts]: That’s not correct. You can definitely use a variable :-)

File: packages/@aws-cdk/aws-codecommit/lib/pipeline-action.ts (hunk 2 above)
Comment (eladb) [reply packages/@aws-cdk/aws-codecommit/lib/pipeline-action.ts]: Yada yada you can ;-)

File: packages/@aws-cdk/aws-codepipeline/lib/stage.ts (hunk 3 above)
Comment (eladb) [reply packages/@aws-cdk/aws-codepipeline/lib/stage.ts]: Got it

File: packages/@aws-cdk/aws-codecommit/lib/pipeline-action.ts (hunk 1 above)
Comment (skinny85) [reply packages/@aws-cdk/aws-codecommit/lib/pipeline-action.ts]: Can you write out the code for that? Cause I can't see it.

**Comment by skinny85** [comment]
Updated with @eladb's feebdack.

File: packages/@aws-cdk/aws-cloudformation/test/test.pipeline-actions.ts [hunk 5]
```diff
@@ -295,6 +296,14 @@ class StageDouble implements cpapi.IStage, cpapi.IInternalStage {
     throw new Error('Unsupported');
   }
 
+  public get pipelineUniqueId(): string {
```
[the remark this answers is not in the document]
Comment (skinny85) [reply packages/@aws-cdk/aws-cloudformation/test/test.pipeline-actions.ts]: Agreed. Refactored in the newest revision.

File: packages/@aws-cdk/aws-codepipeline/lib/pipeline.ts [hunk 6]
```diff
@@ -43,7 +43,7 @@ export interface PipelineProps {
  *
  * // ... add more stages
  */
-export class Pipeline extends cdk.Construct implements events.IEventRuleTarget {
+export class Pipeline extends cdk.Construct implements actions.IPipeline {
```
Comment (eladb) [inline packages/@aws-cdk/aws-codepipeline/lib/pipeline.ts]: the namespace `actions` is weird. I guess you can all this `api` (😉)

File: packages/@aws-cdk/aws-codepipeline/lib/pipeline.ts [hunk 7]
```diff
@@ -216,6 +216,14 @@ export class Pipeline extends cdk.Construct implements events.IEventRuleTarget {
     return this.stages.length;
   }
 
+  public grantBucketRead(identity?: iam.IPrincipal): void {
+    this.artifactBucket.grantRead(identity);
```
Comment (eladb) [inline packages/@aws-cdk/aws-codepipeline/lib/pipeline.ts]: dependency-wise, would it make sense to just expose `artifactBucket` in `IPipeline`?

File: packages/@aws-cdk/aws-codepipeline/lib/stage.ts [hunk 8]
```diff
@@ -76,7 +75,7 @@ export class Stage extends cdk.Construct implements actions.IStage, actions.IInt
   /**
    * The Pipeline this Stage is a part of.
    */
-  public readonly pipeline: Pipeline;
+  public readonly pipeline: actions.IPipeline;
```
Comment (eladb) [inline packages/@aws-cdk/aws-codepipeline/lib/stage.ts]: rename `actions` to something else

File: packages/@aws-cdk/aws-codepipeline/lib/stage.ts [hunk 9]
```diff
@@ -126,7 +117,7 @@ export class Stage extends cdk.Construct implements actions.IStage, actions.IInt
   }
 
   public onStateChange(name: string, target?: events.IEventRuleTarget, options?: events.EventRuleProps) {
-    const rule = new events.EventRule(this.pipeline, name, options);
+    const rule = new events.EventRule((this.pipeline as Pipeline), name, options);
```
Comment (eladb) [inline packages/@aws-cdk/aws-codepipeline/lib/stage.ts]: `this.pipeline as Pipeline` should just be `this`

File: packages/@aws-cdk/aws-codepipeline/lib/pipeline.ts (hunk 6 above)
Comment (skinny85) [reply packages/@aws-cdk/aws-codepipeline/lib/pipeline.ts]: `cpapi`?

File: packages/@aws-cdk/aws-codepipeline/lib/pipeline.ts (hunk 7 above)
Comment (skinny85) [reply packages/@aws-cdk/aws-codepipeline/lib/pipeline.ts]: Do you think I would bother with these methods, if I could just use `artifactBucket` here? 🙂

You can't expose `artifactBucket` in `IPipeline`, as that would be a circular dependency between the `aws-codepipeline-api` and `aws-s3` modules.

File: packages/@aws-cdk/aws-codepipeline/lib/stage.ts (hunk 9 above)
Comment (skinny85) [reply packages/@aws-cdk/aws-codepipeline/lib/stage.ts]: No, it should not, as `events.EventRule` takes a `cdk.Construct` as its first parameter, while `stage.pipeline` is of type `IPipeline` (we cannot make it of type `Pipeline`, which would make this cast unnecessary, as JSII does not allow implementing an interface property with a subtype of the interface declared type).

Comment (skinny85) [reply packages/@aws-cdk/aws-codepipeline/lib/stage.ts]: Oh, did you mean to change the parent of the event to the Stage instead of Pipeline?

**Comment by skinny85** [comment]
Latest round of changes after @eladb's comments.

File: packages/@aws-cdk/aws-codepipeline/lib/stage.ts (hunk 9 above)
Comment (eladb) [reply packages/@aws-cdk/aws-codepipeline/lib/stage.ts]: Yap. Always pass “this”. It’s a simple rule of thumb :-)

Comment (skinny85) [reply packages/@aws-cdk/aws-codepipeline/lib/stage.ts]: Done in the latest revision.

**Comment by skinny85** [comment]
Noticed that there were actually existing unit tests for the CodeCommit's `pollForSourceChanges` functionality in `test.pipeline.ts`, so re-used those ones and removed the redundant ones in `test.actions.ts`.
